### ISSUE TITLE
Remove deprecated API from twisted.python.reflect

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6689]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)| @multani reported|
|-|-|
|Trac ID|trac#6689|
|Type|enhancement|
|Created|2013-08-21 10:56:37Z|
|Branch|https://github.com/twisted/twisted/tree/reflect-deprecate-6689|

Some parts from ```twisted.python.reflect``` are deprecated since [Twisted 12.1](https://twistedmatrix.com/trac/browser/trunk/twisted/python/reflect.py?rev=39660#L48), some since 
[Twisted 11.0](https://twistedmatrix.com/trac/browser/trunk/twisted/python/reflect.py?rev=39660#L373)

As part of porting the rest of ```twisted.python.reflect``` to Python 3 (see [[#5929](https://github.com/twisted/twisted/issues/5929)](https://github.com/twisted/twisted/issues/5929)), I propose to first remove those deprecated part, then to merge #5929.

exarkun seemed to [have confirmed that it should be removed, before porting to Python 3](http://twistedmatrix.com/pipermail/twisted-python/2013-May/026927.html).

Attachments:

* [remove-deprecated-reflect-6689-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/759/759365822e5b9c62604c0b896276562e494eaa37/993d084790da13b253b273f0e52b23cf286a4fc5.patch) (19021 bytes) - added by multani on 2013-08-21 10:58:01Z - 
* [remove-deprecated-reflect-6689-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/759/759365822e5b9c62604c0b896276562e494eaa37/0c92fa721af490cd446260a7fdd33ea956edefcc.patch) (19908 bytes) - added by multani on 2013-08-21 12:54:39Z - 
* [remove-deprecated-reflect-6689-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/759/759365822e5b9c62604c0b896276562e494eaa37/b3565cbfe40ce2de5c105db6c6dc26ca16217723.patch) (19386 bytes) - added by multani on 2013-08-29 14:08:08Z - 
* [remove-deprecated-reflect-6689-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/759/759365822e5b9c62604c0b896276562e494eaa37/3503358a815e4c8d726d26821303961683beb1bf.patch) (19386 bytes) - added by multani on 2013-08-29 14:09:27Z - 
* [remove-deprecated-reflect-6689-5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/759/759365822e5b9c62604c0b896276562e494eaa37/fa8f7f549e55d6fcf839f76fa700adc2fcf351e4.patch) (2231 bytes) - added by multani on 2013-08-30 09:50:57Z - On top of reflect-deprecate-6689 [39808]
* [remove-deprecated-reflect-6689-6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/759/759365822e5b9c62604c0b896276562e494eaa37/c3093dadc326f8f1bb51ffe185abd17f68e73043.patch) (2528 bytes) - added by multani on 2013-08-30 09:58:10Z - On top of reflect-deprecate-6689 [39808]
* [remove-deprecated-reflect-6689-7.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/759/759365822e5b9c62604c0b896276562e494eaa37/15ec5dd0a9707722e9fce30a3ee485dde9d80fd9.patch) (2783 bytes) - added by multani on 2013-09-13 16:30:29Z - On top of reflect-deprecate-6689 [39953]

<details><summary>Searchable metadata</summary>

```
trac-id__6689 6689
type__enhancement enhancement
reporter__multani multani
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_reflect_deprecate_6689 branches/reflect-deprecate-6689
branch_author__multani multani
status__closed closed
resolution__fixed fixed
component__core core
keywords__py3k py3k
time__1377082597000000 1377082597000000
changetime__1379536490000000 1379536490000000
version__None None
owner__thijs thijs
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_2"></a>|@multani commented|
|-|-|

I added the topfile for the removal of the deprecated functions, which I forgot in the first patch.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra set owner to @multani|
|-|-|

This looks good to me but can you get rid of the 'twisted.python.reflect' and ', deprecated since Twisted 12.1.0, has been removed.' repetition in the news file?

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_4"></a>|@multani commented|
|-|-|

Replying to [thijs](#note_3):
> This looks good to me but can you get rid of the 'twisted.python.reflect' and ', deprecated since Twisted 12.1.0, has been removed.' repetition in the news file?

Waht would be a good way to write this? I gave a look at other .removal files and came up with the following:

```
The following attributes in twisted.python.reflect, deprecated since 12.1.0,
have been removed:
    * Accessor
    * AccessorType
    * OriginalAccessor
    * PropertyAccessor
    * Settable
    * Summer

Additionally, the following attributes deprecated since 11.0.0 have also
been removed:
    * accumulateBases
    * allYourBase
```

Would it be better?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

News fragments can't include any special formatting like this.  The newlines and extra spaces will all be removed.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

I was thinking about something like:

```
Accessor, AccessorType, accumulateBases, [etc] in twisted.python.reflect, deprecated since Twisted 12.1.0, are removed now.
```

I wouldn't add a second sentence because of the fact that accumulateBases and allYourBase are deprecated since 11.0 instead of 12.1 but just lump them in one sentence and state they're all deprecated since 12.1.0.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_7"></a>|@multani commented|
|-|-|

I changed the topfile content, as suggested by thijs.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

(In [39807]) Branching to 'reflect-deprecate-6689'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

(In [39808]) apply remove-deprecated-reflect-6689-4.patch. refs [#6689](https://github.com/twisted/twisted/issues/6689)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

I added the patch in a branch but after running `bin/trial twisted.python` the following tests fail:

 * twisted.python.test.test_release.APIBuilderTestCase.test_build
 * twisted.python.test.test_release.APIBuilderTestCase.test_buildWithPolicy

with these exceptions:

```
  File "/Divmod/trunk/Nevow/nevow/flat/ten.py", line 14, in <module>
    from nevow import util
  File "/Divmod/trunk/Nevow/nevow/util.py", line 150, in <module>
    from twisted.python.reflect import qual, namedAny, allYourBase, accumulateBases
exceptions.ImportError: cannot import name allYourBase
```

```
 File "/Divmod/trunk/Nevow/nevow/flat/__init__.py", line 7, in <module>
    from nevow.flat.ten import flatten, precompile, iterflatten, getSerializer, serialize, registerFlattener, getFlattener, partialflatten
  File "/Divmod/trunk/Nevow/nevow/flat/ten.py", line 12, in <module>
    from nevow import inevow
exceptions.ImportError: cannot import name inevow
```

So it looks like these dependencies need to be fixed first. Once they're fixed I imagine we have to mandate a minimum version of pydoctor and/or nevow to use?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

Make sure to install pydoctor when running the tests (otherwise they'll be skipped). Looks like mwhudson worked on a port (https://bazaar.launchpad.net/~mwhudson/pydoctor/dev/revision/598) but this hasn't been merged yet (?). Tested with pydoctor 0.3 and 0.5b1.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

Here's some feedback from IRC:

```
<thijstriemstra> looks like pydoctor still uses nevow
<thijstriemstra> mwhudson: are you going to merge/release https://bazaar.launchpad.net/~mwhudson/pydoctor/dev/revision/598?
<foom> It sounds to me like this removal should be delayed for another 6 months, which remaining uses get sorted out.
<thijstriemstra> true but its been delayed since 11.0 already
<thijstriemstra> we want to get rid of it to make the port of t.p.reflect to python3 possible or easier
<foom> You could put it in "if not python3:"
<foom> (until it's removed)
<radix> that sounds reasonable to me.
```

So opening ticket for pydoctor and nevow (and possibly supplying patches there), and adding a "if not python3:" for `accumulateBases` and `allYourBase` seems like the way forward now. Please make future patches against the ticket's branch, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_13"></a>|@multani commented|
|-|-|

I opened a merge request into Nevow to fix this (see https://code.launchpad.net/~multani/divmod.org/remove-deprecated-twisted.python.reflect/+merge/182937 )

AFAIK, none of these functions/classes are being used by Pydoctor (tested with 0.5b1)

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_14"></a>|@multani commented|
|-|-|

I readded ```accumulateBases``` and ```allYourBase``` to fix the problem found in comment:10.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

Replying to [multani](#note_13):
> I opened a merge request into Nevow to fix this (see https://code.launchpad.net/~multani/divmod.org/remove-deprecated-twisted.python.reflect/+merge/182937 )

Thanks, this has been [merged](https://bazaar.launchpad.net/~divmod-dev/divmod.org/trunk/revision/2716) as well.
 
> AFAIK, none of these functions/classes are being used by Pydoctor (tested with 0.5b1)

Yep.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra commented|
|-|-|

(In [39953]) apply remove-deprecated-reflect-6689-6.patch, refs [#6689](https://github.com/twisted/twisted/issues/6689)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra commented|
|-|-|

Thanks for the patch, I applied it in the branch.

Replying to [multani](#note_14):
> I readded ```accumulateBases``` and ```allYourBase``` to fix the problem found in comment:10.

But the unit tests and `__all__` entries are still missing, can you re-add those as well?

Now that the nevow patch was merged I also suggest changing the comment:

```
#!diff
Index: twisted/python/reflect.py
===================================================================
--- twisted/python/reflect.py	(revision 39953)
+++ twisted/python/reflect.py	(working copy)
@@ -132,12 +132,12 @@
 ## the following were factored out of usage
 
 if not _PY3:
-    # These functions are still imported by libraries used in turn by Twisted,
-    # like Nevow 0.10. Since they are deprecated, there's no need to port them
-    # to Python 3 (hence the condition above).
-    # https://code.launchpad.net/~multani/divmod.org/remove-deprecated-twisted.python.reflect
-    # removes the dependency in Nevow. Once this gets merged and released, these
-    # functions can be safely removed from Twisted.
+    # These functions are still imported by libraries used in turn by the
+    # Twisted unit tests, like Nevow 0.10. Since they are deprecated,
+    # there's no need to port them to Python 3 (hence the condition above).
+    # https://bazaar.launchpad.net/~divmod-dev/divmod.org/trunk/revision/2716
+    # removed the dependency in Nevow. Once this released, these functions can
+    # be safely removed from Twisted.
 
     @deprecated(Version("Twisted", 11, 0, 0), "inspect.getmro")
     def allYourBase(classObj, baseClass=None):
```

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_18"></a>|@multani commented|
|-|-|

Replying to [thijs](#note_17):
> But the unit tests and `__all__` entries are still missing, can you re-add those as well?

Done ...

> Now that the nevow patch was merged I also suggest changing the comment:

... and done.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_19"></a>|@thijstriemstra commented|
|-|-|

(In [40030]) apply remove-deprecated-reflect-6689-7.patch, refs [#6689](https://github.com/twisted/twisted/issues/6689)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

Thanks for the patch multani. A build was [forced](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/reflect-deprecate-6689) and I'll merge when it's green.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_21"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [40032]) Merge reflect-deprecate-6689: Remove deprecated classes from
twisted.python.reflect.

Author: multani
Reviewer: thijs, exarkun
Fixes: [#6689](https://github.com/twisted/twisted/issues/6689)
Refs: [#5929](https://github.com/twisted/twisted/issues/5929)

Accessor, AccessorType, OriginalAccessor, PropertyAccessor,
Settable and Summer in twisted.python.reflect, deprecated
since Twisted 12.1.0, are now removed.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_22"></a>|@thijstriemstra commented|
|-|-|

This needs to be reverted :-/

```
[ERROR]
Traceback (most recent call last):
  File "/home/thijs/workspaces/opensource/software/twisted/svn/Twisted/trunk/twisted/test/test_reflect.py", line 202, in test_accumulateBases
    (Version("Twisted", 11, 0, 0), "inspect.getmro"),
exceptions.NameError: global name 'Version' is not defined

twisted.test.test_reflect.DeprecationTestCase.test_accumulateBases
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/thijs/workspaces/opensource/software/twisted/svn/Twisted/trunk/twisted/test/test_reflect.py", line 192, in test_allYourBase
    (Version("Twisted", 11, 0, 0), "inspect.getmro"),
exceptions.NameError: global name 'Version' is not defined
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_23"></a>|@thijstriemstra set status to `reopened`|
|-|-|

(In [40034]) Revert r40032: Merge reflect-deprecate-6689: Remove deprecated classes from twisted.python.reflect. 

Failing tests because the Version import was removed mistakenly.

[ERROR]
Traceback (most recent call last):
  File "/twisted/test/test_reflect.py", line 202, in test_accumulateBases
    (Version("Twisted", 11, 0, 0), "inspect.getmro"),
exceptions.NameError: global name 'Version' is not defined

twisted.test.test_reflect.DeprecationTestCase.test_accumulateBases
===============================================================================

Reopens: [#6689](https://github.com/twisted/twisted/issues/6689)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_24"></a>|@thijstriemstra set status to `new`|
|-|-|

Sorry about that. Will fix the import and put it back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_25"></a>|@thijstriemstra removed owner|
|-|-|

pyflakes [reported it was safe to remove the import](http://buildbot.twistedmatrix.com/builders/pyflakes/builds/1008/steps/pyflakes/logs/new%20pyflakes%20errors) in test_reflect, probably because it's used like this in `test_allYourBase` and `test_accumulateBases`:

```
self.callDeprecated(
            (Version("Twisted", 11, 0, 0), "inspect.getmro"),
            reflect.allYourBase, DeprecationTestCase)
```

Anyway, forced a new [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/reflect-deprecate-6689), back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

It looks like that pyflakes build was against an older version of the code where `Version` really wasn't used (github link because buildbot won't tell me an svn revision number): https://github.com/twisted/twisted/blob/6c1b18a764f28e1f47633266ef8f701d47ebc91d/twisted/test/test_reflect.py

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_27"></a>|@multani set owner to @thijstriemstra|
|-|-|

Replying to [exarkun](#note_26):
> It looks like that pyflakes build was against an older version of the code where `Version` really wasn't used (github link because buildbot won't tell me an svn revision number): https://github.com/twisted/twisted/blob/6c1b18a764f28e1f47633266ef8f701d47ebc91d/twisted/test/test_reflect.py

I guess the pyflakes build was against revision ```-6``` of the patch, before I added back the tests which required ```Version``` again.

I tested the latest revision of the branch, and I would say it looks good to merge. The 
[http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/reflect-deprecate-6689|Buildbot] looks green, beside a new pyflakes warning on something I didn't change, and the failure on Windows which are due because of lack of space on the disk.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_28"></a>|@hawkowl commented|
|-|-|

The pyflakes warning is because the line number of an existing warning has changed, and can be safely ignored. It shifted the line number of the warning, which it currently counts as a 'new warning'. I'm looking into ways to stop this in future versions of TwistedChecker to prevent this in future.

I had to rerun the Windows build twice to get past the no space error, but it's now green across the board!

Looks good to merge to me.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_29"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [40041]) Re-merge reflect-deprecate-6689: Remove deprecated classes from twisted.python.reflect.

Author: multani
Reviewer: thijs, exarkun, hawkowl
Fixes: [#6689](https://github.com/twisted/twisted/issues/6689)
Refs: [#5929](https://github.com/twisted/twisted/issues/5929)

Accessor, AccessorType, OriginalAccessor, PropertyAccessor, Settable and Summer
in twisted.python.reflect, deprecated since Twisted 12.1.0, are now removed.
